### PR TITLE
Added Deploy Box Product Pathway to GitHub Release and NuGet Release Pipelines

### PR DESCRIPTION
**Pull request by jshigetomi**
# PR Summary

This pull request includes several changes to the `.pipelines` directory, specifically in the `PowerShell-Release-Official.yml`, `release-githubtasks.yml`, `release-publish-nuget.yml`, and `uploadToAzure.yml` files. The changes focus on modifying pipeline stages, adding new stages, and updating tasks to improve the release process.

### Pipeline Configuration Changes:

* Added new stages `PushGitTag` and `MakeDraftPublic` in `PowerShell-Release-Official.yml` to push git tags and make GitHub release drafts public.
* Introduced variables `ob_release_environment` for `msixbundle` and `PublishGitHubRelease` stages to specify the release environment. [[1]](diffhunk://#diff-b25fb4a8cfad5e36ba1c9c9c7cbe494fce4eed0067878610ebb12ffcab7e41cdR121-R122) [[2]](diffhunk://#diff-b25fb4a8cfad5e36ba1c9c9c7cbe494fce4eed0067878610ebb12ffcab7e41cdR284-R315)

### Task and Job Updates:

* Replaced `pwsh` tasks with `PowerShell@2` tasks in `release-githubtasks.yml` to streamline the installation and verification of PowerShell 7, and to clone repositories and set release versions. [[1]](diffhunk://#diff-399a6bedf8e2e8bd47ca1f1f0687a77870f8556867401ef327d44783595092e2L6-R12) [[2]](diffhunk://#diff-399a6bedf8e2e8bd47ca1f1f0687a77870f8556867401ef327d44783595092e2L29-R105) [[3]](diffhunk://#diff-399a6bedf8e2e8bd47ca1f1f0687a77870f8556867401ef327d44783595092e2L106-R145) [[4]](diffhunk://#diff-399a6bedf8e2e8bd47ca1f1f0687a77870f8556867401ef327d44783595092e2L127-R159) [[5]](diffhunk://#diff-399a6bedf8e2e8bd47ca1f1f0687a77870f8556867401ef327d44783595092e2L147-L164)
* Updated `release-publish-nuget.yml` to use `PowerShell@2` tasks for installing PowerShell 7, checking installation, and managing repository checkouts. [[1]](diffhunk://#diff-fe4ea9895cd5ee1821c6ea572b2b9eec7f5fa24cd86c7e7133120ef5b193eeafL11-R17) [[2]](diffhunk://#diff-fe4ea9895cd5ee1821c6ea572b2b9eec7f5fa24cd86c7e7133120ef5b193eeafL32-R104)

### Artifact Management:

* Added steps in `uploadToAzure.yml` to copy GitHub release files and NuGet packages to specific output directories for use in the release pipeline.

These changes aim to enhance the efficiency and reliability of the release process by adding new stages, updating tasks, and improving artifact management.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): 
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
  - [x] N/A or can only be tested interactively
  - **OR**
  - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: 
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: 
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: 
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed:

### CODE REVIEW

File: .pipelines/PowerShell-Release-Official.yml [hunk 1]
```diff
@@ -79,12 +79,18 @@ resources:
 extends:
   template: v2/OneBranch.Official.CrossPlat.yml@templates
   parameters:
+    release: 
+      category: NonAzure
     # still using KS2 because we are not yet using a Box Product Deployment
+    # featureFlags:
```
Comment (TravisEz13) [inline .pipelines/PowerShell-Release-Official.yml:85]: remove dead code please or comment about future work

File: .pipelines/templates/release-githubtasks.yml [hunk 2]
```diff
@@ -26,139 +32,149 @@ jobs:
...
+      targetType: inline
+      pwsh: true
+      script: |
+        Write-Verbose -Verbose "Deploy Box Product Pathway Does Not Support the `"checkout`" task"
+        git clone https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/PowerShell '$(Pipeline.Workspace)/PowerShell'
+        cd $(Pipeline.Workspace)/PowerShell
+        $branchParts = $ENV:BUILD_SOURCEBRANCH -split '/'
+        $branchName = "$($branchParts[-2])/$($branchParts[-1])"
+        git checkout $branchName
+    displayName: Checkout repository
```
Comment (TravisEz13) [inline .pipelines/templates/release-githubtasks.yml:80]:
```suggestion
    displayName: Checkout PowerShell repository
```

File: .pipelines/templates/release-githubtasks.yml [hunk 3]
```diff
@@ -26,139 +32,149 @@ jobs:
...
+      script: |
+        Get-ChildItem Env: | Out-String -width 9999 -Stream | write-Verbose -Verbose
+    displayName: 'Capture Environment Variables'
+
+  - task: PowerShell@2
+    inputs:
+      targetType: inline
+      pwsh: true
+      script: |
+        git clone https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/Internal-PowerShellTeam-Tools '$(Pipeline.Workspace)/tools'
```
Comment (TravisEz13) [inline .pipelines/templates/release-githubtasks.yml:97]: We don't need the repo history.
```suggestion
        git clone --depth 1 https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/Internal-PowerShellTeam-Tools '$(Pipeline.Workspace)/tools'
```

File: .pipelines/templates/release-githubtasks.yml [hunk 4]
```diff
@@ -26,139 +32,149 @@ jobs:
...
-        $packagesPath = Get-ChildItem -Path $destinationPath\*.deb -Recurse -File | Select-Object -First 1 -ExpandProperty DirectoryName
-        Write-Host "sending -- vso[task.setvariable variable=PackagesRoot]$packagesPath"
-        Write-Host "##vso[task.setvariable variable=PackagesRoot]$packagesPath"
+  - task: PowerShell@2
+    inputs:
+      targetType: inline
+      pwsh: true
+      script: |
+        Write-Verbose -Verbose "Deploy Box Product Pathway Does Not Support the `"checkout`" task"
+        git clone https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/PowerShell '$(Pipeline.Workspace)/PowerShell'
```
Comment (TravisEz13) [inline .pipelines/templates/release-githubtasks.yml:75]:
```suggestion
        git clone --branch $(Build.SourceBranch)  --depth  https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/PowerShell '$(Pipeline.Workspace)/PowerShell'
```
I don't think we use the history here either

Comment (TravisEz13) [reply .pipelines/templates/release-githubtasks.yml:75]: I update my recommendation to add the checkout to the clone, which will also speed up the clone.

Comment (TravisEz13) [reply .pipelines/templates/release-githubtasks.yml:75]: Can we make the changelog a build job with an artifact output and update this so we don't need the powershell repo?

File: .pipelines/templates/release-publish-nuget.yml [hunk 5]
```diff
@@ -29,34 +35,73 @@ jobs:
...
+    inputs:
+      targetType: inline
+      pwsh: true
+      script: |
+        Write-Verbose -Verbose "Pwsh 7 Installed"
+        Write-Verbose -Verbose "env:Path: "
+        $env:PATH -split ';' | ForEach-Object {
+          Write-Verbose -Verbose $_
+        }
+    displayName: Check pwsh 7 installation
```
Comment (TravisEz13) [inline .pipelines/templates/release-publish-nuget.yml:70]: make this a step template if you use it multiple times

File: .pipelines/templates/release-publish-nuget.yml [hunk 6]
```diff
@@ -29,34 +35,73 @@ jobs:
...
+      targetType: inline
+      pwsh: true
+      script: |
+        Write-Verbose -Verbose "Deploy Box Product Pathway Does Not Support the `"checkout`" task"
+        git clone https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/PowerShell '$(Pipeline.Workspace)/PowerShell'
+        cd $(Pipeline.Workspace)/PowerShell
+        $branchParts = $ENV:BUILD_SOURCEBRANCH -split '/'
+        $branchName = "$($branchParts[-2])/$($branchParts[-1])"
+        git checkout $branchName
+    displayName: Checkout repository
```
Comment (TravisEz13) [inline .pipelines/templates/release-publish-nuget.yml:83]: previous comment that can simplify this, recommend moving it to a template so changes can be done in one place.

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: .pipelines/PowerShell-Release-Official.yml (hunk 1 above)
Comment (jshigetomi) [reply .pipelines/PowerShell-Release-Official.yml:85]: removed.

File: .pipelines/templates/release-githubtasks.yml (hunk 3 above)
Comment (jshigetomi) [reply .pipelines/templates/release-githubtasks.yml:97]: added

File: .pipelines/templates/release-publish-nuget.yml (hunk 5 above)
Comment (jshigetomi) [reply .pipelines/templates/release-publish-nuget.yml:70]: templatized

File: .pipelines/templates/release-publish-nuget.yml (hunk 6 above)
Comment (jshigetomi) [reply .pipelines/templates/release-publish-nuget.yml:83]: templatized

File: .pipelines/templates/release-githubtasks.yml (hunk 2 above)
Comment (jshigetomi) [reply .pipelines/templates/release-githubtasks.yml:80]: added

File: .pipelines/templates/release-checkout-pwsh-repo.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+steps:
+  - pwsh: |
+      Write-Verbose -Verbose "Deploy Box Product Pathway Does Not Support the `"checkout`" task"
+      git clone --depth 1 https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/PowerShell '$(Pipeline.Workspace)/PowerShell'
```
Comment (TravisEz13) [inline .pipelines/templates/release-checkout-pwsh-repo.yml:4]: depth should be a parameter

File: .pipelines/templates/release-githubtasks.yml (hunk 4 above)
Comment (jshigetomi) [reply .pipelines/templates/release-githubtasks.yml]: We can make the CL and Tools clone an artifact output in a separate stage.

File: .pipelines/templates/release-checkout-pwsh-repo.yml (hunk 7 above)
Comment (jshigetomi) [reply .pipelines/templates/release-checkout-pwsh-repo.yml:4]: If we make a separate stage for CL and tools upload, then I'm seeing this template as unnecessary...

File: .pipelines/PowerShell-Release-Official.yml [hunk 8]
```diff
@@ -270,15 +270,49 @@ extends:
         parameters:
           SkipPSInfraInstallers: ${{ parameters.SkipPSInfraInstallers }}
 
+    - stage: UploadChangelogAndTools
```
Comment (adityapatwardhan) [inline .pipelines/PowerShell-Release-Official.yml:273]: Remove this for now

File: .pipelines/PowerShell-Release-Official.yml [hunk 9]
```diff
@@ -270,15 +270,49 @@ extends:
...
+    - stage: MakeDraftPublic
+      displayName: Make Draft Public
+      dependsOn: PushGitTag
+      jobs:
+      - template: /.pipelines/templates/approvalJob.yml@self
+        parameters:
+          displayName: Make Draft Public
+          jobName: DraftPublic
+          instructions: |
+            Make the GitHub Release Draft Public
```
Comment (adityapatwardhan) [inline .pipelines/PowerShell-Release-Official.yml:309]: These can be combined into a single stage for push and make draft public as both are approval jobs

File: .pipelines/templates/release-githubtasks.yml [hunk 10]
```diff
@@ -26,139 +32,106 @@ jobs:
...
-      Write-Verbose -Verbose -Message $fileContent
+  - task: PowerShell@2
+    inputs:
+      targetType: inline
+      pwsh: true
+      script: |
+        $Path = "$(Pipeline.Workspace)/GitHubPackages"
+        $OutputPath = Join-Path $Path 'hashes.sha256'
+        $srcPaths = @($Path)
+        $packages  = Get-ChildItem -Path $srcPaths -Include * -Recurse -File
```
Comment (adityapatwardhan) [inline .pipelines/templates/release-githubtasks.yml:65]: We can just use $Path
```suggestion
        $packages  = Get-ChildItem -Path $Path -Include * -Recurse -File
```

File: .pipelines/templates/release-githubtasks.yml [hunk 11]
```diff
@@ -26,139 +32,106 @@ jobs:
...
-      $fileContent = Get-Content -Path $OutputPath -Raw | Out-String
-      Write-Verbose -Verbose -Message $fileContent
+  - task: PowerShell@2
+    inputs:
+      targetType: inline
+      pwsh: true
+      script: |
+        $Path = "$(Pipeline.Workspace)/GitHubPackages"
+        $OutputPath = Join-Path $Path 'hashes.sha256'
+        $srcPaths = @($Path)
```
Comment (adityapatwardhan) [inline .pipelines/templates/release-githubtasks.yml:64]: Remove this

File: .pipelines/templates/uploadToAzure.yml [hunk 12]
```diff
@@ -233,6 +235,24 @@ jobs:
       Get-ChildItem '$(Build.ArtifactStagingDirectory)/downloads' | Select-Object -ExpandProperty FullName
     displayName: 'Capture downloads'
 
+  - pwsh: |
+      Write-Verbose -Verbose "Copying Github Release files in $(Build.ArtifactStagingDirectory)/downloads to use in Release Pipeline"
+      
+      Write-Verbose -Verbose "Creating output directory for GitHub Release files: $(ob_outputDirectory)/GitHubPackages"
+      New-Item -Path $(ob_outputDirectory)/GitHubPackages -ItemType Directory -Force
+      Get-ChildItem -Path "$(Build.ArtifactStagingDirectory)/downloads/*" -Recurse | 
+        Where-Object { $_.Extension -notin '.msix', '.nupkg' } |
```
Comment (adityapatwardhan) [inline .pipelines/templates/uploadToAzure.yml:244]: add in `.msixbundle`

File: .pipelines/templates/release-publish-nuget.yml [hunk 13]
```diff
@@ -29,34 +35,30 @@ jobs:
     value: $(Build.SourcesDirectory)\PowerShell\.config\suppress.json
 
   steps:
-  - checkout: self
-    clean: true
-    env:
-      ob_restore_phase: true # This ensures checkout is done at the beginning of the restore phase
+  - template: release-install-pwsh.yml
```
Comment (adityapatwardhan) [inline .pipelines/templates/release-publish-nuget.yml:38]: Add an approval job before this

**Review by adityapatwardhan — APPROVED** [review]

File: .pipelines/templates/release-checkout-pwsh-repo.yml [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+      Write-Verbose -Verbose "Deploy Box Product Pathway Does Not Support the `"checkout`" task"
+      if ($ENV:BUILD_REASON -eq 'PullRequest') {
+        throw 'We dont support PRs'
+      }
+      
+      Write-Verbose -Verbose $ENV:BUILD_SOURCEBRANCH
+      $branchName = $ENV:BUILD_SOURCEBRANCH -replace '^refs/heads/'
+      Write-Verbose -Verbose "Branch Name: $branchName"
+      git clone --depth 1 --branch $branchName https://$(mscodehubCodeReadPat)@mscodehub.visualstudio.com/PowerShellCore/_git/PowerShell '$(Pipeline.Workspace)/PowerShell'
+      cd $(Pipeline.Workspace)/PowerShell
```
Comment (TravisEz13) [inline .pipelines/templates/release-checkout-pwsh-repo.yml:12]: It won't remember this, but not blocking

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**